### ISSUE TITLE
Multiple assertions with the same key

### ISSUE THREAD

**Issue by fatlotus** [russellhaering/gosaml2#9]
Howdy,

I noticed that currently assertions are stored as a `map[string]string` — meaning that two assertions with the same key (i.e. a multi-valued LDAP property) will get overwritten. I thought of two ways to fix this:

1. Break compatibility, and switch to a `map[string][]string`, possibly wrapped like a `url.Form`, so that the common case is simple.
2. Keep the `AssertionInfo` map, but add a supplementary list of `Assertion` objects. (This would also allow some implementations to use the friendly attribute name.)

Thoughts?

**Comment by russellhaering** [comment]
I like the `url.Form` pattern and I'm open to breaking compatibility.

@andrewstuart since I get the sense you're using this, any thoughts?

**Comment by andrewstuart** [comment]
Yeah, that's a pattern used all over the place in Go (also in url.Values, http.Header, e.g.), so it makes sense to me, and wouldn't cause me any issues. I might try to throw together a quick PR for that as well tonight, to have a custom type like the net/* packages are fond of doing.

**Comment by fatlotus** [comment]
If so, can I request a way of getting (or looking up by) the FriendlyName field as well? I'm seeing behavior somewhat like the following (this is for an Internet2 Shib IdP):

```xml
<saml2:Attribute
  FriendlyName="eduPersonEntitlement"
  Name="urn:oid:1.3.6.1.4.1.5923.1.1.1.7"
  NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
  <saml2:AttributeValue
      xmlns:xsd="http://www.w3.org/2001/XMLSchema"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:type="xsd:string">
    urn:mace:dir:entitlement:common-lib-terms
  </saml2:AttributeValue>
  <saml2:AttributeValue
      xmlns:xsd="http://www.w3.org/2001/XMLSchema"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:type="xsd:string">
    https://university.edu/path/to/thingy
  </saml2:AttributeValue>
  <saml2:AttributeValue
      xmlns:xsd="http://www.w3.org/2001/XMLSchema"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:type="xsd:string">
    urn:mace:incommon:entitlement:common:1
  </saml2:AttributeValue>
</saml2:Attribute>
```

**Comment by andrewstuart** [comment]
@fatlotus I'm all for that too, as the OIDs are pretty hard to work with. I will note, however that [the spec](https://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf) (line 1201) explicitly mentions that "This attribute's value MUST NOT be used as a basis for formally identifying SAML attributes."

So we may want to consider what such an API might look like, such `UnsafeGetFriendlyName(fname)` or some such method name. Or maybe a strong note in the docs would be enough. Thoughts?

**Comment by fatlotus** [comment]
Oof. Perhaps a bit obtuse, but maybe override `.String()` to return the raw XML? That way a human can `.Printf(...)` the result for hints, but would be strongly discouraged from hard-coding a FriendlyName into the software.

**Comment by andrewstuart** [comment]
So I've got something pulled together [here](/andrewstuart/gosaml2/tree/assertion-multivalue/values.go), but this method (since it uses type aliasing rather than a struct) would be unable to store any additional metadata like FriendlyName.

Another alternative, perhaps more flexible, is to just create an Attribute struct, and use map[string][]Attribute for AssertionInfo.Values:

```go
type Attribute struct {
  Values []string
  Name string
  FriendlyName string
}
```

This has the benefit of being multi-valued, including FriendlyName, yet still being indexed by Name, hopefully somewhat discouraging reliance upon FriendlyName as the key.

**Comment by russellhaering** [comment]
I like that. I'd combine it with the type alias like you've already got for easy access by Name.

**Comment by andrewstuart** [comment]
Okay, I put something together that I think has a pretty clean API. Let me know what you think. :-)

**Comment by fatlotus** [comment]
Closed by #10.
